### PR TITLE
[add] GTree based Decision Tree

### LINKED ISSUE #212: 使用 SPU 实现决策树模型基础功能
> 此 ISSUE 为 [隐语开源共建计划（SecretFlow Open Source Contribution Plan，简称 SF OSCP）第二期](https://github.com/orgs/secretflow/discussions/741)任务 ISSUE，欢迎社区开发者参与共建～
> 若有感兴趣想要认领的任务，但还未报名，辛苦先完成[报名](https://www.wjx.top/vm/QKio0dq.aspx# )进行哈～

## 任务介绍

- 任务名称：使用 SPU 实现决策树模型基础功能
- 技术方向：SPU/SML
- 任务难度：进阶🌟🌟

## 详细要求：
- 安全性（尽量少 reveal）
- 功能性：实现算法的基本功能，包括：
  - 需要支持最基本的训练和预测
  - 起码一种 criterion，如 gini
  - 支持一些简单的预剪枝机制或其他防止过拟合的手段
- 收敛性：包含 simulator 跑出的实验数据并且证明收敛性
- 代码规范：Python 代码需要使用 black+isort 进行格式化（流水线包含代码规范检查卡点）
- 提交说明：关联该 isuue 并提交代码至 https://github.com/secretflow/spu/tree/main/sml
- 特殊说明：若某个特性有特殊的限制，如需要 FM128，需要更多 fxp 等需要在注释文档中明确说明
- 任务难点：该任务几乎都是比较运算，密态下每个子节点的数据量理论上都是 O(n)，可能导致效率很低，需要在实现功能的基础上尽可能提高效率

## 能力要求：
- 熟悉经典的机器学习算法
- 熟悉 JAX 或 NumPy，可以使用 NumPy 实现算法

## 操作说明：
- 可参考明文实现 [DecisionTreeClassifier](https://github.com/scikit-learn/scikit-learn/blob/364c77e04/sklearn/tree/_classes.py#L595)
- 操作指引：https://www.secretflow.org.cn/docs/spu/latest/en-US/getting_started/tutorials/develop_your_first_mpc_application
- 范例：https://github.com/secretflow/spu/blob/main/sml/linear_model/simple_sgd.py

### PR DESCRIPTION
**Pull request by ElleryQu**
# Pull Request

## What problem does this PR solve?

Issue Number: Fixed #212 

## Possible side effects?

- Performance:

**test**:

```
No GPU/TPU found, falling back to CPU. (Set TF_CPP_MIN_LOG_LEVEL=0 and rerun for more info.)
[2023-10-13 18:32:00.502] [info] [thread_pool.cc:30] Create a fixed thread pool with size 19
Accuracy in SKlearn: 0.96
Accuracy in SPU: 0.96
.
----------------------------------------------------------------------
Ran 1 test in 78.026s

OK
```

**emulation**:

Running time in SPU: 77.37s
Accuracy in SKlearn: 0.96
Accuracy in SPU: 0.96
[2023-10-13 18:35:38,091] Shutdown multiprocess cluster...

- Backward compatibility:

### CODE REVIEW

**Comment by ElleryQu** [comment]
I have read the CLA Document and I hereby sign the CLA

File: sml/tree/emulations/tree_emul.py [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+        n_labels = len(jnp.unique(iris_label))
+        sorted_features = jnp.sort(iris_data, axis=0)
+        new_threshold = (sorted_features[:-1, :] + sorted_features[1:, :]) / 2
+        new_features = (
+            jnp.greater_equal(iris_data[:, :], new_threshold[:, jnp.newaxis, :]) + 1 - 1
+        )
+        new_features = new_features.transpose([1, 0, 2]).reshape(n_samples, -1)
+
+        X, y = new_features, iris_label
+        return X, y
```
Comment (deadlywing) [inline sml/tree/emulations/tree_emul.py:52]: 我发现特征爆炸有点严重，测试时间太久了；
建议改成`X[:50, ::3], y[:50]`以减少总的测试的时间

File: sml/tree/tests/tree_test.py [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+            new_threshold = (sorted_features[:-1, :] + sorted_features[1:, :]) / 2
+            new_features = (
+                jnp.greater_equal(iris_data[:, :], new_threshold[:, jnp.newaxis, :])
+                + 1
+                - 1
+            )
+            new_features = new_features.transpose([1, 0, 2]).reshape(n_samples, -1)
+
+            X, y = new_features, iris_label
+            return X, y
```
Comment (deadlywing) [inline sml/tree/tests/tree_test.py:56]: 同emulation，，减少一下sample数和feature数

File: sml/tree/tree.py [hunk 3]
```diff
@@ -0,0 +1,270 @@
...
+# limitations under the License.
+
+
+import jax.numpy as jnp
+
+
+
+class DecisionTreeClassifier:
+    """A decision tree classifier.
+
```
Comment (deadlywing) [inline sml/tree/tree.py:22]: 1. 可以贴一下Gtree的paper链接，并简单介绍一下Gtree的整体思路；
2. 需要着重说明一下模型对数据的要求：如data需要事先分箱（即0,1），支持多分类，但label必须是0,1,2...这种格式
3. 能提供一下训练和预测的复杂度量级么？（这个的目的主要是想让用户对这个算法有一个大概的性能感知，以及了解哪些参数会影响复杂度）

File: sml/tree/tree.py [hunk 4]
```diff
@@ -0,0 +1,270 @@
...
+        # (n_array, n_index)
+        O = array[:, jnp.newaxis] * E  # select shares
+        zu = jnp.sum(O, axis=0)
+    # OAA vectorization variant
+    elif len(array.shape) == 2:
+        # (n_arrays, n_array, n_index)
+        O = array[:, :, jnp.newaxis] * E[jnp.newaxis, :, :]  # select shares
+        zu = jnp.sum(O, axis=1)
+    else:
+        raise "Not implemented."
```
Comment (deadlywing) [inline sml/tree/tree.py:83]: 同上

File: sml/tree/tree.py [hunk 5]
```diff
@@ -0,0 +1,270 @@
...
+    new_gamma = new_gamma.at[0::2, :].set(gamma)
+    new_gamma = new_gamma.at[1::2, :].set(gamma)
+
+    # # modification.
+    psi = jnp.zeros((n_leaves, n_labels))
+    # print(psi.shape)
+    for i in range(n_labels):
+        psi = psi.at[:, i].set(Cn[:, i + 1, 0] + Cn[:, i + 1, 1])
+    total = jnp.sum(psi, axis=1)
+    # print(total.shape)
```
Comment (deadlywing) [inline sml/tree/tree.py:188]: 这里的print语句可以删一下哈，，

**Comment by deadlywing** [comment]
@ElleryQu 抱歉，，前段时间有点忙，delay了一段时间；
整体实现应该ok，，有一些小的建议；

另外，麻烦下次push之前对代码的格式做一些调整：
1. 使用 buildifier 格式化bazel文件
2. 使用black和isort 格式化python文件

**Comment by ElleryQu** [comment]
> @ElleryQu 抱歉，，前段时间有点忙，delay了一段时间； 整体实现应该ok，，有一些小的建议；
> 
> 另外，麻烦下次push之前对代码的格式做一些调整：
> 
> 1. 使用 buildifier 格式化bazel文件
> 2. 使用black和isort 格式化python文件

好嘞，感谢

**Comment by ElleryQu** [comment]
> @ElleryQu 抱歉，，前段时间有点忙，delay了一段时间； 整体实现应该ok，，有一些小的建议；
> 
> 另外，麻烦下次push之前对代码的格式做一些调整：
> 
> 1. 使用 buildifier 格式化bazel文件
> 2. 使用black和isort 格式化python文件

您好，上述问题已经修改。
1. 对于特征爆炸的问题：修改为`X[:, ::3], y[:]`。
2. 在DecisionTreeBuilder类增加了一段描述性文字。原始文章未给出复杂度分析，这里附上了一个大概的估算。
3. 根据上一次检查结果进行了格式化处理。对于其他杂项进行了修改。

File: sml/tree/tests/tree_test.py [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+            iris = load_iris()
+            iris_data, iris_label = jnp.array(iris.data), jnp.array(iris.target)
+            # sorted_features: n_samples * n_features_in
+            n_samples, n_features_in = iris_data.shape
+            n_labels = len(jnp.unique(iris_label))
+            sorted_features = jnp.sort(iris_data, axis=0)
+            new_threshold = (sorted_features[:-1, :] + sorted_features[1:, :]) / 2
+            new_features = (
+                jnp.greater_equal(iris_data[:, :], new_threshold[:, jnp.newaxis, :])
+                + 1
```
Comment (deadlywing) [inline sml/tree/tests/tree_test.py:51]: 这里有点confusing，，+1-1不是抵消了么？

File: sml/tree/tree.py [hunk 7]
```diff
@@ -0,0 +1,273 @@
...
+        O = array[:, :, jnp.newaxis] * E[jnp.newaxis, :, :]  # select shares
+        zu = jnp.sum(O, axis=1)
+    return zu
+
+
+def oaa_elementwise(array, index_array):
+    '''
+    Extract elements from each sub-array of array according to index_array.
+
+    e.g. array = [[1, 2, 3], [4, 5, 6]], index_array = [0, 2], output = [[1, 3], [4, 6]].
```
Comment (deadlywing) [inline sml/tree/tree.py:105]: 这里似乎有点问题？我感觉output应该是[1,6]啊？

File: sml/tree/emulations/tree_emul.py [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+    def load_data():
+        iris = load_iris()
+        iris_data, iris_label = jnp.array(iris.data), jnp.array(iris.target)
+        # sorted_features: n_samples * n_features_in
+        n_samples, n_features_in = iris_data.shape
+        n_labels = len(jnp.unique(iris_label))
+        sorted_features = jnp.sort(iris_data, axis=0)
+        new_threshold = (sorted_features[:-1, :] + sorted_features[1:, :]) / 2
+        new_features = (
+            jnp.greater_equal(iris_data[:, :], new_threshold[:, jnp.newaxis, :]) + 1 - 1
```
Comment (deadlywing) [inline sml/tree/emulations/tree_emul.py:50]: 同test

**Comment by deadlywing** [comment]
python format 仍然有点问题

File: sml/tree/tests/tree_test.py (hunk 6 above)
Comment (deadlywing) [reply sml/tree/tests/tree_test.py:51]: 这里似乎还没改

**Comment by deadlywing** [comment]
test文件中 +1-1那里还没有修改哈～

**Comment by ElleryQu** [comment]
> test文件中 +1-1那里还没有修改哈～

您好，这里是为了保证jax.equal结果在SPU上是加法共享，如果改成+0会被解释器优化掉。

如果不加+1-1，jax.equal的结果为布尔分享，赋值时会发生错误。review展示了一个最小的错误例子。这种错误会导致在iris_data上分类正确率降低至66%。

**Comment by deadlywing** [comment]
> > test文件中 +1-1那里还没有修改哈～
> 
> 您好，这里是为了保证jax.equal结果在SPU上是加法共享，如果改成+0会被解释器优化掉。
> 
> 如果不加+1-1，jax.equal的结果为布尔分享，赋值时会发生错误。review展示了一个最小的错误例子。这种错误会导致在iris_data上分类正确率降低至66%。

hi，我自己本地试了一下，并没有发现准确性下降的情况，，能辛苦贴一下具体的运行setting和结果么？

**Comment by ElleryQu** [comment]
> > > test文件中 +1-1那里还没有修改哈～
> > 
> > 
> > 您好，这里是为了保证jax.equal结果在SPU上是加法共享，如果改成+0会被解释器优化掉。
> > 如果不加+1-1，jax.equal的结果为布尔分享，赋值时会发生错误。review展示了一个最小的错误例子。这种错误会导致在iris_data上分类正确率降低至66%。
> 
> hi，我自己本地试了一下，并没有发现准确性下降的情况，，能辛苦贴一下具体的运行setting和结果么？

没有复现出来，但之前确实遇到过这个问题(￣ ￣|| )

删除了+1-1

**Review by deadlywing — APPROVED** [review]
LGTM
